### PR TITLE
Fix and refactor `deprecated` decorator to `deprecate_func`

### LINKED ISSUE #6585: Inconsistent warning types for deprecations
Currently our decorators that are used in deprecation use different warning types:

- [`remove_arg` raises `FutureWarning`](https://github.com/scikit-image/scikit-image/blob/c3cfd5fca4180883a0e304e1616424c8d4cc14d9/skimage/_shared/utils.py#L155)
- [`deprecated_kwarg` raises `FutureWarning`](https://github.com/scikit-image/scikit-image/blob/c3cfd5fca4180883a0e304e1616424c8d4cc14d9/skimage/_shared/utils.py#L276)
- [`deprecated` raises `skimage_deprecation`](https://github.com/scikit-image/scikit-image/blob/c3cfd5fca4180883a0e304e1616424c8d4cc14d9/skimage/_shared/utils.py#L473)

Is there a reason to use different types? Otherwise I'd consider this an unnecessary for our users and an API inconsistency.

### LINKED ISSUE #6586: deprecated decorator warning does not point at calling location
### Description:

Contrary to our other decorators such as `deprecate_kwarg`, `deprecated` points at the source code of the decorated function and not at the calling location which would be more useful to the user:

https://github.com/scikit-image/scikit-image/blob/c3cfd5fca4180883a0e304e1616424c8d4cc14d9/skimage/_shared/utils.py#L470-L475

I think these decorators could use some cleanup and love in general. See also #6585. I'm happy to look into this, but want to get the other WIP for the release done before.

### Way to reproduce:

```Python
import numpy as np
from skimage.filters.lpi_filter import inverse

def filt_func(r, c, sigma = 1):
    return np.exp(-np.hypot(r, c)/sigma)

inverse(np.ones((3, 3)), filt_func)
```

### Traceback or output:

```Shell
scikit-image/skimage/filters/lpi_filter.py:171: skimage_deprecation: Function ``inverse`` is deprecated and will be removed in version 0.21. Use ``skimage.filters.lpi_filter.filter_inverse`` instead.
  removed_version='0.21')
Out[2]: 
array([[0.78918759, 0.41273506, 0.66250022],
       [0.41273506, 0.03946844, 0.37050594],
       [0.66250022, 0.37050594, 0.95194294]])
```

### Version information:

```Shell
3.10.6 | packaged by conda-forge | (main, Aug 22 2022, 20:35:26) [GCC 10.4.0]
Linux-6.0.2-arch1-1-x86_64-with-glibc2.36
scikit-image version: 0.20.0.dev0+git20221016.c3157562e
numpy version: 1.23.3
```

### PR DESCRIPTION
**Pull request by lagru**
## Description

- Closes #6585. `skimage_deprecation` was only used in the decorator `deprecated` while everywhere else FutureWarning is used. So let's be consistent about this. 
- Also don't modify the warning filter when calling the wrapped function. That this happens is very implicit and hidden from the user. I'd argue that it's the users responsibility to configure his warning filter while it's our responsibility to chose good and expected defaults.
- Closes #6586. Previously, the warning did not point to where the deprecated function was called (user code) but at the deprecated function's definition (library code)! Add a test for this.
- I also propose an updated signature of the decorator, that makes sure that the deprecated version is specified (forgotten in previous uses of this decorator). I replaced `alt_func` with a more general `hint` parameter as there is not always a clear alternative function. In that case it's possible to provide a more flexible guiding message to the user. It also uses the imperative mood like other similar decorators.
- Also do away with the behavior parameter. Raising an exception defeats the purpose of a deprecation decorator as it immediately breaks user code. It impossible to use the deprecated function, in which case it should just be removed.
- Update the current uses of the decorator.

## Checklist

- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- Gallery example in `./doc/examples` (new features only)
- Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- Unit tests
- Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- Descriptive commit messages (see below)

## For reviewers

- Check that the PR title is short, concise, and will make sense 1 year
  later.
- Check that new functions are imported in corresponding `__init__.py`.
- Check that new features, API changes, and deprecations are mentioned in
  `doc/release/release_dev.rst`.
- There is a bot to help automate backporting a PR to an older branch. For
  example, to backport to v0.19.x after merging, add the following in a PR
  comment: `@meeseeksdev backport to v0.19.x`
- To run benchmarks on a PR, add the `run-benchmark` label. To rerun, the label
  can be removed and then added again. The benchmark output can be checked in
  the "Actions" tab.

### CODE REVIEW

**Comment by stefanv** [comment]
Thanks for the refactor, @lagru.

I wonder if it's worth capturing the version in which a function was deprecated; does that matter much to the user? We didn't have that before.

**Comment by lagru** [comment]
Yes, I think it's useful to provide the warning when something was deprecated. It makes it easier to find related resources (e.g. release notes) and increases the specificity of the search message. That's useful to users and to us during user support and maintenance as well. It might even prevent errors such as the on reverted in https://github.com/scikit-image/scikit-image/pull/6583/commits/998ee00833194a3fab29d0816737a309d382ab84. Also, this pattern is already used in our [`deprecate_kwarg` decorator](https://github.com/scikit-image/scikit-image/blob/e0e7ed7fecf33986553c9b6ca362a36bd9c4fcc1/skimage/_shared/utils.py#L237-L238).

**Comment by stefanv** [comment]
OK, in that case I'd use the wording "X was deprecated in skimage v0.19, and will be removed in v0.22"

**Comment by mkcor** [comment]
> "X was deprecated in skimage v0.19, and will be removed in v0.22"

Just to be on the same page, shouldn't it be "X was deprecated in skimage v0.19, and will be removed in **v0.21**" in the general case (deprecation cycle of two minor releases)?

File: skimage/_shared/tests/test_utils.py [hunk 1]
```diff
@@ -300,3 +300,27 @@ def test_decorator_warnings():
 
     assert record[0].lineno == expected_lineno
     assert record[0].filename == __file__
+
+
+@deprecate_func(
```
Comment (mkcor) [inline skimage/_shared/tests/test_utils.py:305]:
```suggestion
@deprecated_func(
```
I'm not sure about my suggestion: I think that *formally* it helps with consistency (`deprecated_func` replaced `deprecated`, plus the dummy function below is named `_deprecated_func`), but maybe you intended that and you are absolutely logical (the decorator *deprecates* while the dummy function stands for a *deprecated* function).

File: skimage/_shared/tests/test_utils.py [hunk 2]
```diff
@@ -300,3 +300,27 @@ def test_decorator_warnings():
...
+)
+def _deprecated_func():
+    """Dummy function used in `test_deprecate_func`.
+
+    The decorated function must be outside the test function, otherwise it
+    seems that the warning does not point at the calling location.
+    """
+
+
+def test_deprecate_func():
```
Comment (mkcor) [inline skimage/_shared/tests/test_utils.py:316]:
```suggestion
def test_deprecated_func():
```

File: skimage/_shared/utils.py [hunk 3]
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
         return fixed_func
 
 
-class deprecated:
-    """Decorator to mark deprecated functions with warning.
+class deprecate_func(_DecoratorBaseClass):
+    """Decorator a deprecated functions and warn when it is called.
```
Comment (mkcor) [inline skimage/_shared/utils.py:427]:
```suggestion
    """Decorate a deprecated function and warn when it is called.
```

File: skimage/_shared/utils.py [hunk 4]
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
...
-    behavior : {'warn', 'raise'}
-        Behavior during call to deprecated function: 'warn' = warn user that
-        function is deprecated; 'raise' = raise error.
+    deprecated_version : str
+        The package version during which the deprecation was introduced.
     removed_version : str
         The package version in which the deprecated function will be removed.
+    hint : str, optional
+        A hint on how to address this deprecation.
+        E.g. "Use `skimage.submodule.alternative_func` instead."
```
Comment (mkcor) [inline skimage/_shared/utils.py:439]:
```suggestion
        A hint on how to address this deprecation,
        e.g., "Use `skimage.submodule.alternative_func` instead."
```

File: skimage/_shared/utils.py [hunk 5]
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
...
 
     Parameters
     ----------
-    alt_func : str
-        If given, tell user what function to use instead.
-    behavior : {'warn', 'raise'}
-        Behavior during call to deprecated function: 'warn' = warn user that
-        function is deprecated; 'raise' = raise error.
+    deprecated_version : str
+        The package version during which the deprecation was introduced.
```
Comment (mkcor) [inline skimage/_shared/utils.py:434]:
```suggestion
        The package version when the deprecation was introduced.
```

File: skimage/feature/__init__.py [hunk 6]
```diff
@@ -30,15 +30,17 @@
 from ._basic_features import multiscale_basic_features
 
 
-@deprecated(alt_func='skimage.feature.graycomatrix',
-            removed_version='1.0')
+@deprecate_func(hint='Use `skimage.feature.graycomatrix` instead.',
+                deprecated_version="0.19",
```
Comment (mkcor) [inline skimage/feature/__init__.py:34]:
```suggestion
                deprecated_version='0.19',
```

File: skimage/feature/__init__.py [hunk 7]
```diff
@@ -30,15 +30,17 @@
...
+                removed_version='1.0')
 def greycomatrix(image, distances, angles, levels=None, symmetric=False,
                  normed=False):
     return graycomatrix(image, distances, angles, levels, symmetric, normed)
 
 
-@deprecated(alt_func='skimage.feature.graycoprops',
-            removed_version='1.0')
+@deprecate_func(hint='Use `skimage.feature.graycoprops` instead.',
+                deprecated_version="0.19",
```
Comment (mkcor) [inline skimage/feature/__init__.py:42]:
```suggestion
                deprecated_version='0.19',
```

File: skimage/_shared/utils.py [hunk 8]
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
...
-                raise skimage_deprecation(msg)
+            stacklevel = 1 + self.get_stack_length(func) - stack_rank
+            warnings.warn(
+                message,
+                category=FutureWarning,
+                stacklevel=stacklevel
+            )
             return func(*args, **kwargs)
 
         # modify doc string to display deprecation warning
```
Comment (mkcor) [inline skimage/_shared/utils.py:470]:
```suggestion
        # modify docstring to display deprecation warning
```

File: skimage/_shared/utils.py [hunk 9]
```diff
@@ -432,53 +423,52 @@ def fixed_func(*args, **kwargs):
...
+        self.deprecated_version=deprecated_version
         self.removed_version = removed_version
+        self.hint = hint
 
     def __call__(self, func):
+        message = (
+            f"`{func.__name__}` is deprecated since version "
+            f"{self.deprecated_version}"
+        )
+        if self.deprecated_version:
```
Comment (grlee77) [inline skimage/_shared/utils.py:452]: typo?
```suggestion
        if self.removed_version:
```

File: skimage/_shared/tests/test_utils.py (hunk 1 above)
Comment (grlee77) [reply skimage/_shared/tests/test_utils.py:305]: It seems okay to leave as `deprecate_func`, but no strong opinion either way

**Review by grlee77 — COMMENTED** [review]
There are some comments and conflicts to resolve, but it looks good to me once that is done

File: skimage/_shared/tests/test_utils.py (hunk 1 above)
Comment (mkcor) [reply skimage/_shared/tests/test_utils.py:305]: @grlee77 ok, thanks.
@lagru feel free to ignore this suggestion.

File: skimage/_shared/utils.py (hunk 9 above)
Comment (mkcor) [reply skimage/_shared/utils.py:452]: Looks like it, good catch!

**Comment by lagru** [comment]
Thanks for the reviews both of you! I addressed your comments and resolved the conflicts. I think the [CI failure](https://github.com/scikit-image/scikit-image/actions/runs/3637875098/jobs/6139398577#step:6:623) is unrelated so this should be good to go from my side.

**Comment by lagru** [comment]
Maybe we can still get this bug fix into 0.20. :pray:

**Comment by stefanv** [comment]
This is fine with me, if you want to pick it back up.

File: skimage/filters/ridges.py [hunk 10]
```diff
@@ -14,12 +14,17 @@ class of ridge filters relies on the eigenvalues of the Hessian matrix of
...
-from .._shared.utils import _supported_float_type, check_nD, deprecated
+from .._shared.utils import _supported_float_type, check_nD, deprecate_func
 from ..feature.corner import hessian_matrix, hessian_matrix_eigvals
 from ..util import img_as_float
 
 
-@deprecated(alt_func='skimage.feature.hessian_matrix', removed_version="0.21")
+@deprecate_func(
+    deprecated_version="0.20",
+    removed_version="0.21",
```
Comment (jarrodmillman) [inline skimage/filters/ridges.py:24]:
```suggestion
    removed_version="0.22",
```

Comment (jarrodmillman) [reply skimage/filters/ridges.py:24]: Also we should add `compute_hessian_eigenvalues` to `TODO.txt`.

**Review by jarrodmillman — APPROVED** [review]
LGTM (w/ the suggested changes).

**Review by stefanv — APPROVED** [review]

**Comment by stefanv** [comment]
Thanks, Lars!

**Pull request merged**